feat: Group synced emotes

CritteRo opened the pull request:
This PR adds the ability to start group synced emotes for all emotes except SHARED ones.

Features:
* Using `/gemote [emotename]` or `LCTRL + Enter` in the emote menu allows you to start a group emote at your location, for everyone within 5 units of distance.
* Emote requestor automatically accepts the group request.
* The rest of the people have 10 seconds to press Y to accept the group emote.
* Leaving the 5-unit radius area, or starting another emote, will cancel the group emote for you.
* At the end of the 10 seconds wait (or if everyone in the radius accepts the emote), everyone who accepted the emote and is still in the radius, will start the emote.

Fixes:
* Fixed `:AddInstructionButton` in NativeUI, to be able to actually use the method.

Limitations:
* Current implementation is hard-set to only use 5 units, with the Requestor Player as the origin point, for the group radius. Ideally, we should have a dynamic range, that calculates if a player is in range of another player that is in range of the origin (to allow for things like conga lines, lol)
* The system does the initial distance check for eligible players client-side, and then sends that list server-side for re-checking. I did this to not put a strain on servers with lots of players. (Imagine doing 100+ distance checks server-side every time someone started a group emote). This might need to be changed if we want to address the previous limitation.
* We use the regular `OnEmotePlay()` method to start the group anim, but we don't check if the animation is actually available for the player when we request it. So, animal peds might still receive and accept group emotes, but then won't be able to play them. Doesn't break anything, but it's a bad user experience.

CritteRo commented:
Fixes (#104 )

Manason commented on server/GroupEmote.lua:
```diff
@@ -0,0 +1,110 @@
+
+    -- Add Players to the group request
+    local sourceCoords = GetEntityCoords(GetPlayerPed(source))
+    for i=1, 128, 1 do -- Max number of players that can be in one scope.
```
Why not just iterate over the array sent by the client?

Manason commented on server/GroupEmote.lua:
```diff
@@ -0,0 +1,110 @@
+            local ped = GetPlayerPed(src)
+            local crds = DoesEntityExist(ped) and GetEntityCoords(ped)
+            if #(sourceCoords - crds) <= groupEmoteRequests[reqid].zoneRadius then
+                groupEmoteRequests[reqid].players[src] = false
```
Maybe this should be nil instead of false? You'll be able to simplify some later code to avoid checking for a true value, iterating over the players will be enough to confirm they are in the set.

Manason commented on server/GroupEmote.lua:
```diff
@@ -0,0 +1,110 @@
+    group.players[source] = true
+
+    local startGroupEmote = true
+    for i,k in pairs(group.players) do
```
use more descriptive variable names

Manason commented on client/GroupEmote.lua:
```diff
@@ -0,0 +1,123 @@
+AddEventHandler("rpemotes:client:autoCancel", function()
+    while groupEmoteReqId ~= nil and groupEmoteAccepted == true do
+        -- Keep rechecking to make sure the player is still in area.
+        local crds = GetEntityCoords(PlayerPedId())
```
Feels like we should just check this once server side before sending out the trigger for the group to start emoting, although it's nice that with the client check, they get notified they aren't in the group.

CritteRo replied:
We do it server-side as well, when the emote starts. I added this client-side for player convenience.

CritteRo replied:
We do it server-side as well, when the emote starts. I added this because clients with a pending group emote request (accepted or not), cannot start their own group emote request, and cannot receive other group emote requests.

Manason commented on client/GroupEmote.lua:
```diff
@@ -0,0 +1,123 @@
+            SimpleNotify(Translate('refusedgroupemote'))
+        end
+
+        if IsControlJustPressed(1, 246) then
```
Shouldn't this be user bindable?

CritteRo commented on server/GroupEmote.lua:
```diff
@@ -0,0 +1,110 @@
+        if not players[i] or type(players[i]) ~= "number" then
+            break
+        end
+        if GetPlayerPing(players[i]) ~= 0 then
```
GetPlayerPing() returns 0 if the param is not a player, or the player is offline. We make sure the player is valid and online, before moving forward with the logic.

CritteRo replied:
GetPlayerPing() returns 0 if the param is not a player, or the player is offline. We make sure the player is valid and online, before moving forward with the logic. It should not cause more network traffic, as the data is already gathered and cached by the server. Should just read from the same place as https://docs.fivem.net/natives/?_0x9A928294

CritteRo commented on server/GroupEmote.lua:
```diff
@@ -0,0 +1,110 @@
+    -- Create groupEmoteRequest
+    ::retry_reqid::
+    local reqid = tostring(os.time())..tostring(source)..tostring(math.random(0,99))
+    if groupEmoteRequests[reqid] then goto retry_reqid end
```
While we do keep track of who requested the group emote, they are not necessarily the owner of that emote request, just another player that happens to be the first one to accept it. 

This means that a player can (and in my opinion should be able to) start a group emote for one set of players, but cancel the emote themselves to go start another one. This is why I didn't block the group emote creation server-side, but I did block it client-side. Each client (requestor or not) can only have one active group emote request (not canceled or rejected) at one time.

This allows the client to not be blocked by a group emote request for 10 seconds (or longer, if the server decides to increase the time-to-respond).

Manason commented on client/GroupEmote.lua:
```diff
@@ -52,15 +55,19 @@ RegisterNetEvent("rpemotes:client:startGroupEmote", function(reqid, zone)
+        local currentEmote = LocalPlayer.state.currentEmote
         local crds = GetEntityCoords(PlayerPedId())
-        if #(crds - (groupEmoteOriginCoords or vector3(0))) > groupEmoteOriginRadius or LocalPlayer.state.currentEmote ~= nil then
+        if #(crds - (groupEmoteOriginCoords or vector3(0))) > groupEmoteOriginRadius or (currentEmote ~= nil and currentEmote ~= existingEmote) then
```
On second thought, I think we should allow players to emote in the meantime but then be forced into the group emote. I'm imagining a scenario where players are dancing. It would be a bit weird if we forced them to sit there for 10 seconds and do nothing in the meantime. I think it's better if they can still do other emotes, but just can't leave the area.

CritteRo commented:
@Manason please review.

* Added ability for players to select the group radius of the emote.
* Added config vars for the default, and max group radius.

* Moved `helpText()` function from `Placement.lua` to `Utils.lua`, and made it using labels (because "STRING" limits the text to 99 characters).
* Fixed conflicts with main branch.
